### PR TITLE
Termainate declarations#2542

### PR DESCRIPTION
**Pull request by marinakr**

### CODE REVIEW

File: config/test.exs [hunk 1]
```diff
@@ -31,6 +31,9 @@ config :ops, OPS.EventManagerRepo,
   pool: Ecto.Adapters.SQL.Sandbox,
   ownership_timeout: 120_000_000
 
+config :ops, OPS.DeclarationsAutoProcessor,
+  termination_batch_size: {:system, :integer, "DECLARATION_REQUEST_AUTOTERMINATION_BATCH", 1}
```
Comment (AlexKovalevych) [inline config/test.exs]: you may set it directly to `1` if not planning to change the value dynamically in the tests

File: lib/ops/declaration/declaration_autoprocessor.ex [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        send(pid, :autoterminate)
+
+      error ->
+        Logger.error("Declarations was not terminated: #{inspect(error)}")
+    end
+
+    {:noreply, state}
+  end
+
+  def declaration_processor(gen_server) do
```
Comment (AlexKovalevych) [inline lib/ops/declaration/declaration_autoprocessor.ex]: may have a better name - `set_processor_state` or `init_processor_state`, etc

File: lib/ops/declaration/declaration_autoprocessor.ex [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      Declaration
+      |> select([d], %{id: d.id})
+      |> where([d], d.inserted_at < datetime_add(^NaiveDateTime.utc_now(), ^(-1 * expiration), ^unit))
+      |> where([d], d.status == ^Declaration.status(:pending))
+      |> limit(^limit)
+
+    updates = [status: Declaration.status(:active), updated_by: user_id, updated_at: DateTime.utc_now()]
+
+    query =
+      Declaration
```
Comment (AlexKovalevych) [inline lib/ops/declaration/declaration_autoprocessor.ex]: no need to use pipe if you have a single function call

File: lib/ops/declaration/declaration_autoprocessor.ex [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      |> select([d], %{id: d.id})
+      |> where([d], d.end_date < ^NaiveDateTime.utc_now())
+      |> where([d], d.status not in ^[Declaration.status(:closed), Declaration.status(:terminated)])
+      |> limit(^limit)
+
+    updates = [status: Declaration.status(:closed), updated_by: user_id, updated_at: DateTime.utc_now()]
+
+    query =
+      Declaration
```
Comment (AlexKovalevych) [inline lib/ops/declaration/declaration_autoprocessor.ex]: same here

File: lib/ops/declaration/declaration_autoprocessor.ex [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      |> join(:inner, [d], dr in subquery(subquery), dr.id == d.id)
+
+    update_rows(query, updates)
+  end
+
+  defp update_rows(query, updates) do
+    Repo.transaction(fn ->
+      updates = Repo.update_all(query, [set: updates], returning: Declarations.updated_fields_list(updates))
```
Comment (AlexKovalevych) [inline lib/ops/declaration/declaration_autoprocessor.ex]:
```
{rows_updated, rows} = Repo.update_all...
Declarations.log_status_updates(rows)
rows_updated
```

File: priv/repo/migrations/20180704156722_add_declaration_status_end_date_index.exs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+defmodule OPS.Repo.Migrations.AddDeclarationStatusEndDateIndex do
+  use Ecto.Migration
+
+  @disable_ddl_transaction true
+  def change do
+    create(index(:declarations, [:status, :end_date], concurrently: true))
```
Comment (AlexKovalevych) [inline priv/repo/migrations/20180704156722_add_declaration_status_end_date_index.exs]: we already have such index

File: test/integration/declaration_auto_approve_test.exs [hunk 7]
```diff
@@ -3,21 +3,28 @@ defmodule OPS.Integration.DeclarationAutoApproveTest do
...
   alias OPS.Declarations.Declaration
+  alias OPS.DeclarationsAutoProcessor
   alias OPS.Repo
 
   test "approve_declarations/0" do
-    declaration = insert(:declaration, status: Declaration.status(:pending))
-    inserted_at = NaiveDateTime.add(NaiveDateTime.utc_now(), -86_400 * 10, :seconds)
+    declaration_ids =
+      Enum.map(1..10, fn _ ->
+        declaration = insert(:declaration, status: Declaration.status(:pending))
```
Comment (AlexKovalevych) [inline test/integration/declaration_auto_approve_test.exs]: you can set inserted_at here

File: test/integration/declaration_terminator_test.exs [hunk 8]
```diff
@@ -3,21 +3,28 @@ defmodule OPS.DeclarationTerminatorTest do
...
   alias OPS.Declarations.Declaration
+  alias OPS.DeclarationsAutoProcessor
   alias OPS.Repo
 
   test "terminate_declarations/0" do
-    declaration = insert(:declaration, status: Declaration.status(:pending))
-    inserted_at = NaiveDateTime.add(NaiveDateTime.utc_now(), -86_400 * 10, :seconds)
+    declaration_ids =
+      Enum.map(1..10, fn _ ->
+        declaration = insert(:declaration, status: Declaration.status(:pending))
```
Comment (AlexKovalevych) [inline test/integration/declaration_terminator_test.exs]: same here

File: priv/repo/migrations/20180704156722_add_declaration_status_end_date_index.exs (hunk 6 above)
Comment (marinakr) [reply priv/repo/migrations/20180704156722_add_declaration_status_end_date_index.exs]: to do: return this index
remove old similar index

**Pull request merged**